libabi/Abi.cpp

```diff
@@ -0,0 +1,521 @@
+		type = ABIType(_arg.substr(0, _arg.find(':')));
+		val = _arg.substr(_arg.find(':') + 1);
+	}
+	else
+		val = _arg;
+	
+	if (_prefix != Tristate::False)
+	{
+		if (val.substr(0, 2) == "0x")
+		{
+			type.noteHexInput(val.size() - 2);
+			return make_tuple(fromHex(val), type, Format::Hex);
+		}
+		if (val.substr(0, 1) == "+")
+		{
+			type.noteDecimalInput();
+			return make_tuple(toCompactBigEndian(bigint(val.substr(1))), type, Format::Decimal);
+		}
+		if (val.substr(0, 1) == "'")
+		{
+			type.noteBinaryInput();
+			return make_tuple(asBytes(val.substr(1)), type, Format::Binary);
+		}
+		if (val == "[")
+			return make_tuple(bytes(), type, Format::Open);
+		if (val == "]")
+			return make_tuple(bytes(), type, Format::Close);
+	}
+	if (_prefix != Tristate::True)
+	{
+		if (val.find_first_not_of("0123456789") == string::npos)
+		{
+			type.noteDecimalInput();
+			return make_tuple(toCompactBigEndian(bigint(val)), type, Format::Decimal);
+		}
+		if (val.find_first_not_of("0123456789abcdefABCDEF") == string::npos)
+		{
+			type.noteHexInput(val.size());
+			return make_tuple(fromHex(val), type, Format::Hex);
+		}
+		if (val == "[")
+			return make_tuple(bytes(), type, Format::Open);
+		if (val == "]")
+			return make_tuple(bytes(), type, Format::Close);
+		type.noteBinaryInput();
+		return make_tuple(asBytes(val), type, Format::Binary);
+	}
+	throw InvalidUserString();
+}
+
+	
+ABIMethod::ABIMethod(js::mObject _o)
+{
+	name = _o["name"].get_str();
+	isConstant = _o["constant"].get_bool();
+	if (_o.count("inputs"))
+		for (auto const& i: _o["inputs"].get_array())
+		{
+			js::mObject a = i.get_obj();
+			ins.push_back(ABIType(a["type"].get_str(), a["name"].get_str()));
+		}
+	if (_o.count("outputs"))
+		for (auto const& i: _o["outputs"].get_array())
+		{
+			js::mObject a = i.get_obj();
+			outs.push_back(ABIType(a["type"].get_str(), a["name"].get_str()));
+		}
+}
+
+ABIMethod::ABIMethod(string const& _name, vector<ABIType> const& _args)
+{
+	name = _name;
+	ins = _args;
+}
+
+string ABIMethod::sig() const
+{
+	string methodArgs;
+	for (auto const& arg: ins)
+		methodArgs += (methodArgs.empty() ? "" : ",") + arg.canon();
+	return name + "(" + methodArgs + ")";
+}
+FixedHash<4> ABIMethod::id() const { return FixedHash<4>(sha3(sig())); }
+	
+std::string ABIMethod::solidityDeclaration() const
+{
+	ostringstream ss;
+	ss << "function " << name << "(";
+	int f = 0;
+	for (ABIType const& i: ins)
+		ss << (f++ ? ", " : "") << i.canon() << " " << i.name;
+	ss << ") ";
+	if (isConstant)
+		ss << "constant ";
+	if (!outs.empty())
+	{
+		ss << "returns(";
+		f = 0;
+		for (ABIType const& i: outs)
+			ss << (f++ ? ", " : "") << i.canon() << " " << i.name;
+		ss << ")";
+	}
+	return ss.str();
+}
+
+bytes ABIMethod::encode(vector<pair<bytes, Format>> const& _params) const
+{
+	bytes ret = name.empty() ? bytes() : id().asBytes();
+	bytes suffix;
+	
+	// int int[] int
+	// example: 42 [ 1 2 3 ] 69
+	// int[2][][3]
+	// example: [ [ [ 1 2 3 ] [ 4 5 6 ] ] [ ] ]
```

Comment:
Not concerning this PR, but that change should be made once this PR is merged:
`int[2][][3]` should be a 3-array of an array of a 2-array of ints, i.e. the example should be
`[ [ [ 1 2 ] [ 3 4 ] ] [ ] [ ] ]`.
Furthermore, there are plans to change the encoding of nested dynamic arrays (those are not really covered in the specification).